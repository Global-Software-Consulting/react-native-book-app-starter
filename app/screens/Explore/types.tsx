import { Books } from 'models/reducers/appReducers';
import { ReactNodeArray } from 'react';

export interface Props {
    books?: Array<Books>;
    name?: string;
    base_url?: string;
    onRefresh: () => void;
}

export interface IParams {
    id: number;
    averageRating: number;
    title: string;
    numberOfPages: number;
    shortSummary: string;
    item: { id: number };
}
