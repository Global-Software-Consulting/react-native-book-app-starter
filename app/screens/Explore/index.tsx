import { useIsFocused } from '@react-navigation/core';
import SearchBar from 'components/SearchBar';
import { ReducerState } from 'models/reducers/index';
import React, { useEffect, useState } from 'react';
import { Alert, BackHandler, View } from 'react-native';
import { useTheme } from 'react-native-paper';
import { useDispatch, useSelector } from 'react-redux';
import * as appActions from 'store/actions/appActions';
import * as loadingActions from 'store/actions/loginActions';
import ExploreComponent from './Container';
import ExploreShimmer from './Shimmer';
import Config from 'react-native-config';

const Explore: React.FC = () => {
    //fetching book images from the store
    const books = useSelector((state: ReducerState) => state.appReducer.books);
    const isLoading = useSelector((state: ReducerState) => state.loadingReducer.isLoading);
    const isFocused = useIsFocused();
    const userData = useSelector((state: ReducerState) => state.loginReducer.user);
    const dispatch = useDispatch();
    const [mounted, setMounted] = useState(true);
    const theme = useTheme();
    useEffect(() => {
        setMounted(true);
        if (isFocused) {
            if (mounted) fetchBookDetails();
        }
        return () => {
            setMounted(false);
            dispatch(loadingActions.disableLoader());
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isFocused, userData]);

    //handling back hardware button
    useEffect(() => {
        const backAction = () => {
            Alert.alert('Book App', 'Are you sure you want to exit?', [
                {
                    text: 'Cancel',
                    onPress: () => null,
                    style: 'cancel',
                },
                { text: 'YES', onPress: () => BackHandler.exitApp() },
            ]);
            return true;
        };

        const backHandler = BackHandler.addEventListener('hardwareBackPress', backAction);

        return () => {
            backHandler.remove;
        };
    });

    const fetchBookDetails = async () => {
        if (mounted) dispatch(appActions.getBookRequest('a'));
    };

    return (
        <View style={{ backgroundColor: theme.colors.background, flex: 1, paddingHorizontal: 10 }}>
            <SearchBar />
            {isLoading ? (
                <ExploreShimmer />
            ) : (
                <ExploreComponent
                    name={userData?.firstName + ' ' + userData?.lastName}
                    books={books}
                    onRefresh={fetchBookDetails}
                />
            )}
        </View>
    );
};

export default Explore;
