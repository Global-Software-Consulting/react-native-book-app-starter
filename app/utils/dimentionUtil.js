import { useDeviceOrientation } from '@react-native-community/hooks';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { Dimensions, PixelRatio } from 'react-native';

const widthPercentageToDP = (screenWidth, widthPercent) => {
    const elemWidth = typeof widthPercent === 'number' ? widthPercent : parseFloat(widthPercent);
    return PixelRatio.roundToNearestPixel((screenWidth * elemWidth) / 100);
};

const heightPercentageToDP = (screenHeight, heightPercent) => {
    const elemHeight =
        typeof heightPercent === 'number' ? heightPercent : parseFloat(heightPercent);
    return PixelRatio.roundToNearestPixel((screenHeight * elemHeight) / 100);
};

export function getPercentageWidth() {
    const [width, setWidth] = useState(Dimensions.get('window').width);
    useEffect(() => {
        Dimensions.addEventListener('change', (newDimensions) => {
            // Retrieve and save new dimensions
            screenWidth = newDimensions.window.width;
            setWidth(screenWidth);
        });
        return Dimensions.removeEventListener('change', () => {});
    }, []);
    const wp = useCallback((value) => widthPercentageToDP(width, value), [width]);

    return wp;
}
export function getPercentageHeight() {
    const [height, setHeight] = useState(Dimensions.get('window').height);
    useEffect(() => {
        Dimensions.addEventListener('change', (newDimensions) => {
            screenHeight = newDimensions.window.height;
            setHeight(screenHeight);
        });
        return Dimensions.removeEventListener('change', () => {});
    }, []);
    const hp = useCallback((value) => heightPercentageToDP(height, value), [height]);

    return hp;
}
