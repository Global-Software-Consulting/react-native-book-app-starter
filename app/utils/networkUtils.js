import NetInfo from "@react-native-community/netinfo";
export default class NetworkUtils {
  static async isNetworkAvailable() {
    const response = await NetInfo.fetch();
    return response.isConnected;
}}