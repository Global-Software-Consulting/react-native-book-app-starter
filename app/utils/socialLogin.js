import auth from '@react-native-firebase/auth';
import { LoginManager, AccessToken } from 'react-native-fbsdk';
import { GoogleSignin } from '@react-native-google-signin/google-signin';


export default class socialLogin {

static async loginWithGoogle() {
// Get the users ID token
const { idToken } = await GoogleSignin.signIn();

// Create a Google credential with the token
const googleCredential = auth.GoogleAuthProvider.credential(idToken);

// Sign-in the user with the credential
return auth().signInWithCredential(googleCredential);
}

static async loginWithFacebook() {
        // Attempt login with permissions
  const result = await LoginManager.logInWithPermissions(['public_profile', 'email']);

  if (result.isCancelled) {
    throw 'User cancelled the login process';
  }

  // Once signed in, get the users AccesToken
  const data = await AccessToken.getCurrentAccessToken();

  if (!data) {
    throw 'Something went wrong obtaining access token';
  }

  // Create a Firebase credential with the AccessToken
  const facebookCredential = auth.FacebookAuthProvider.credential(data.accessToken);

  // Sign-in the user with the credential
  return auth().signInWithCredential(facebookCredential);
}

}
