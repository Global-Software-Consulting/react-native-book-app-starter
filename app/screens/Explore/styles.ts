import { Platform, StyleSheet } from 'react-native';
import { useTheme } from 'react-native-paper';
import { getPercentageHeight, getPercentageWidth } from 'utils/dimentionUtil';
import { Dimensions } from 'react-native';
export const useStyles = () => {
    const theme = useTheme();
    const width = getPercentageWidth();
    const height = getPercentageHeight();
    const windowWidth = Dimensions.get('window').width;
    const windowHeight = Dimensions.get('window').height;

    const styles = StyleSheet.create({
        container: {
            flex: 1,
            paddingVertical: 10,
            paddingHorizontal: 30,
            backgroundColor: theme.colors.background,
            marginBottom: height('10%'),
        },
        favoriteView: {
            alignItems: 'center',
            justifyContent: 'center',
            marginBottom: height('6%')
        },
        bookmark: {
            margin: 2,
            fontWeight: 'bold',
        },
        imageError: {
            height: windowHeight > windowWidth ? height('50%') : height('100%'),
            width: width('50%'),
            tintColor: theme.colors.text,
        },
        mainViewSetting: { height: height('100%') },
        name: {
            fontSize: windowHeight > windowWidth ? width('5%') : width('3%'),
            fontWeight: 'bold',
            fontFamily: 'Avenir-Medium',
            alignItems: 'center',
            color: theme.colors.text,
        },
        middleView: {
            marginBottom: windowHeight > windowWidth ? height('15%') : Platform.OS==="ios" ? height('25%') :height('40%') ,
        },
        nameShimmer: {
            width: width('25%'),
            height: height('6%'),
            backgroundColor: '#E7E5E7',
            marginBottom: 2,
        },
        tagLine: {
            fontSize: windowHeight > windowWidth ? width('4%') : width('2%'),
            fontFamily: 'Avenir-Medium',
            alignItems: 'center',
            color: theme.colors.text,
        },
        tagLineShimmer: {
            width: width('40%'),
            height: height('4%'),
            backgroundColor: '#E7E5E7',
        },
        listCaption: {
            fontSize: windowHeight > windowWidth ? width('4%') : width('2%'),
            marginTop: height('2%'),
            fontFamily: 'Avenir-Medium',
            alignItems: 'center',
            fontWeight: 'bold',
            color: theme.colors.text,
        },
        listCaptionShimmer: {
            width: width('20%'),
            height: height('3%'),
            backgroundColor: '#E7E5E7',
        },
        button: {
            backgroundColor: '#61e3a5',
            padding: 10,
            borderRadius: 10,
            margin: 10,
        },

        horizontalRuler: {
            borderBottomColor: '#DCDCDC',
            borderBottomWidth: 1,
            marginTop: height('0.5%'),
            color: theme.colors.primary,
        },
        flatList: {
            flexGrow: 1,
            marginTop: 10,
        },
        flatListLast: {
            flexGrow: 1,
            marginTop: 10,
            marginBottom: height('3%'),
        },
        searchView: {
            marginTop: 20,
            marginBottom: 20,
            flexDirection: 'row',
            borderColor: 'grey',
            borderWidth: 1,
            borderRadius: 30,
            justifyContent: 'space-between',
            alignItems: 'center',
        },
        searchViewShimmer: {
            marginBottom: 20,
            flexDirection: 'row',
            borderRadius: 30,
            justifyContent: 'space-between',
            alignItems: 'center',
            height: height('5%'),
            backgroundColor: '#E7E5E7',
        },
        searchViewChildren: {
            margin: 2,
            marginLeft: 5,
            color: theme.colors.text,
        },
        books: {
            height: height('2%'),
            width: width('1%'),
            borderRadius: 4,
        },

        bookTrending: {
            height: height('25%'),
            width: width('37%'),
            borderRadius: 4,
        },

        bookGeneral: {
            height: height('20%'),
            width: width('30%'),

            borderRadius: 4,
        },
        bookView: {
            backgroundColor: 'transparent',
            paddingRight: 20,
        },
    });
    return styles;
};
