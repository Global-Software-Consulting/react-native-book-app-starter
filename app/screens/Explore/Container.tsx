import { useNavigation } from '@react-navigation/native';
//importing card component
import BookCard from 'components/BookCard/BookCard';
import images from 'config/images';
import { Books, IAppState } from 'models/reducers/appReducers';
import React from 'react';
import { useTranslation } from 'react-i18next';
import {
    FlatList,
    Image,
    RefreshControl,
    ScrollView,
    TouchableHighlight,
    View,
} from 'react-native';
import { Text, useTheme } from 'react-native-paper';
import { useSelector } from 'react-redux';
import { useStyles } from 'screens/Explore/styles';
import { Props } from './types';
const ExploreComponent: React.FC<Props> = (props) => {
    //theme handling
    const styles = useStyles();
    const { t } = useTranslation();
    const theme = useTheme();
    const navigation = useNavigation();
    const { name, onRefresh } = props;
    const dummyImages = [
        'https://d1csarkz8obe9u.cloudfront.net/posterpreviews/sad-romance-kindle-book-cover-flyer-template-539b5fdf2dd2f6602c25ce81fbb5d877.jpg?ts=1589326539',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSO4yGwkz128dpVHBztwERbm6Z6kIXwQ03V0A&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQQB-3-cJ6tpQ6cFUjCach0R1dTiXdhXI0AtA&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRuaVjRvjIZAO6XIsd0eHS22Q949ke7r-KZFg&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRkIVlYNBoYdfq1FGWs_dx_k8Irvjmdu64p7A&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRwor6J4qPx1WWN27GYiZEFP-NAoWSstn-YrQ&usqp=CAU',
    ];

    const generateRandomURL = () => {
        const number = Math.random() * 6;
        const index = Math.floor(number);
        return dummyImages[index];
    };
    //getting data from store
    const books = useSelector((state: { appReducer: IAppState }) => state.appReducer.books);
    const isLoading = useSelector(
        (state: { loadingReducer: IAppState }) => state.loadingReducer.isLoading,
    );

    const navigateToDetails = (params: number) => {
        navigation.navigate('BookDetail' as never, params as never);
    };

    const bookList = (item: Books, style: 'General' | 'Custom') => {
        return (
            <TouchableHighlight
                key={item.id}
                underlayColor={theme.colors.highlight}
                onPress={() => {
                    navigateToDetails(item.id);
                }}>
                <BookCard
                    url={generateRandomURL()}
                    styleSelect={style}
                    bookTitle={item?.title}
                    book={item}
                    id={item?.id}
                />
            </TouchableHighlight>
        );
    };

    return (
        <View style={styles.mainViewSetting}>
            <ScrollView
                nestedScrollEnabled={true}
                style={styles.container}
                showsVerticalScrollIndicator={false}
                refreshControl={
                    <RefreshControl
                        refreshing={isLoading}
                        onRefresh={onRefresh}
                        colors={['green', 'yellow']}
                    />
                }>
                <View style={styles.middleView}>
                    <Text style={styles.name}>
                        {t('Hi')} {name}{' '}
                    </Text>
                    <Text style={styles.tagLine}>{t('Lets find something new')}</Text>

                    {books?.length > 0 ? (
                        <View>
                            <View style={styles.horizontalRuler} />
                            <Text style={styles.listCaption}>{t('Trending')}</Text>
                            <FlatList
                                nestedScrollEnabled={true}
                                horizontal
                                data={books?.filter((item: Books) => {
                                    return item.averageRating > 3;
                                })}
                                contentContainerStyle={styles.flatList}
                                renderItem={({ item }) => bookList(item, 'Custom')}
                                showsHorizontalScrollIndicator={false}
                                keyExtractor={(item, index) => 'key' + index}
                            />
                            <View style={styles.horizontalRuler} />
                            <Text style={styles.listCaption}>{t('New Releases')}</Text>
                            <FlatList
                                horizontal
                                data={books?.filter((item: Books) => {
                                    return item.averageRating <= 3 && item.averageRating > 0;
                                })}
                                contentContainerStyle={styles.flatList}
                                renderItem={({ item }) => bookList(item, 'General')}
                                showsHorizontalScrollIndicator={false}
                                keyExtractor={(item, index) => 'key' + index}
                            />
                            <View style={styles.horizontalRuler} />
                            <Text style={styles.listCaption}>{t('Selected for you')}</Text>
                            <FlatList
                                horizontal
                                data={books?.filter((item: Books) => {
                                    return item.averageRating === 0;
                                })}
                                contentContainerStyle={styles.flatListLast}
                                renderItem={({ item }) => bookList(item, 'General')}
                                showsHorizontalScrollIndicator={false}
                                keyExtractor={(item, index) => 'key' + index}
                            />
                        </View>
                    ) : (
                        <View style={styles.favoriteView}>
                            <Image source={images.books.noBookFound} style={styles.imageError} />
                            <Text style={styles.bookmark}>No books found</Text>
                        </View>
                    )}
                </View>
            </ScrollView>
        </View>
    );
};

export default ExploreComponent;
