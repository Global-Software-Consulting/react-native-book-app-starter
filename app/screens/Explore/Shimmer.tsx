//importing card component
import BookCardShimmer from 'components/BookCard/BookCardShimmer';
import React from 'react';
import { FlatList, ScrollView, TouchableHighlight, View } from 'react-native';
import { useTheme } from 'react-native-paper';
import { useStyles } from './styles';
const ExploreShimmer: React.FC = () => {
    const styles = useStyles();
    const theme = useTheme();
    const bookList = (item: number, style: 'General' | 'Custom') => {
        return (
            <TouchableHighlight key={item} underlayColor={theme.colors.highlight}>
                <BookCardShimmer styleSelect={style} />
            </TouchableHighlight>
        );
    };
    return (
        <ScrollView showsVerticalScrollIndicator={false}>
            <View style={{ paddingLeft: 20 }}>
                <View style={styles.nameShimmer} />
                <View style={styles.tagLineShimmer} />

                <View style={styles.horizontalRuler} />

                <View style={styles.listCaptionShimmer} />
                <FlatList
                    horizontal
                    data={[0, 0, 0]}
                    contentContainerStyle={styles.flatList}
                    renderItem={({ item }) => bookList(item, 'Custom')}
                    showsHorizontalScrollIndicator={false}
                    showsVerticalScrollIndicator={false}
                    keyExtractor={(item, index) => 'key' + index}
                />
                <View style={styles.horizontalRuler} />

                <View style={styles.listCaptionShimmer} />
                <FlatList
                    horizontal
                    data={[0, 0, 0]}
                    contentContainerStyle={styles.flatList}
                    renderItem={({ item }) => bookList(item, 'General')}
                    showsHorizontalScrollIndicator={false}
                    showsVerticalScrollIndicator={false}
                    keyExtractor={(item, index) => 'key' + index}
                />

                <View style={styles.horizontalRuler} />

                <View style={styles.listCaptionShimmer} />
                <FlatList
                    horizontal
                    data={[0, 0, 0]}
                    contentContainerStyle={styles.flatListLast}
                    renderItem={({ item }) => bookList(item, 'General')}
                    showsVerticalScrollIndicator={false}
                    showsHorizontalScrollIndicator={false}
                    keyExtractor={(item, index) => 'key' + index}
                />
            </View>
        </ScrollView>
    );
};

export default ExploreShimmer;
