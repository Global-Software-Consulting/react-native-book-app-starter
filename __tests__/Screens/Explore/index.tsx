import { useIsFocused, useNavigation } from '@react-navigation/native'
import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import * as redux from 'react-redux'

import { act } from 'react-test-renderer'
import Explore from 'screens/Explore/index'
import * as utils from 'utils/dimentionUtil'
import { getPercentageHeight, getPercentageWidth } from 'utils/dimentionUtil'
const DeviceTypeUtilsMock = jest.requireMock('utils/dimentionUtil')
import 'react-native-gesture-handler/jestSetup'

// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

jest.mock('./../../../app/utils/dimentionUtil', () => {
    return {
        getPercentageHeight: jest.fn(),
        height: '100',
        getPercentageWidth: jest.fn(),
    }
})

describe('Screen test', () => {
    test('Snapshot with loading as true ', () => {
        const height = jest.fn()
        getPercentageHeight.mockReturnValue(jest.fn())
        const width = jest.fn()
        getPercentageWidth.mockReturnValue(jest.fn())

        const isFocused = jest.fn()
        useIsFocused.mockReturnValue(true)

        act(() => {
            const state = {
                loadingReducer: { isLoading: true },
                loginReducer: { user: 'as' },
                appReducer: {
                    books: [
                        { averageRating: 2 },
                        { averageRating: 4 },
                        { averageRating: 5 },
                        { averageRating: 1 },
                        { averageRating: 3 },
                    ],
                },
            }
            jest.spyOn(redux, 'useSelector').mockImplementation((val) => {
                val(state)
                expect(tree).toMatchSnapshot()
            })
        })
        const tree = render(<Explore />)
    })

    test('Snapshot with loading as false ', () => {
        const height = jest.fn()
        getPercentageHeight.mockReturnValue(jest.fn())
        const width = jest.fn()
        getPercentageWidth.mockReturnValue(jest.fn())

        const isFocused = jest.fn()
        useIsFocused.mockReturnValue(true)

        act(() => {
            const state = {
                loadingReducer: { isLoading: false },
                loginReducer: { user: 'as' },
                appReducer: {
                    books: [
                        { averageRating: 2 },
                        { averageRating: 4 },
                        { averageRating: 5 },
                        { averageRating: 1 },
                        { averageRating: 3 },
                    ],
                },
            }
            jest.spyOn(redux, 'useSelector').mockImplementation((val) => {
                val(state)
                expect(tree).toMatchSnapshot()
            })
        })
        const tree = render(<Explore />)
    })
})
