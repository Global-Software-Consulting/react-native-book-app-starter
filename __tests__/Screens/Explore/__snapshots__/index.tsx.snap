// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen test Snapshot with loading as false  1`] = `undefined`;

exports[`Screen test Snapshot with loading as false  2`] = `undefined`;

exports[`Screen test Snapshot with loading as false  3`] = `undefined`;

exports[`Screen test Snapshot with loading as false  4`] = `undefined`;

exports[`Screen test Snapshot with loading as false  5`] = `undefined`;

exports[`Screen test Snapshot with loading as true  1`] = `undefined`;

exports[`Screen test Snapshot with loading as true  2`] = `undefined`;

exports[`Screen test Snapshot with loading as true  3`] = `undefined`;

exports[`Screen test Snapshot with loading as true  4`] = `undefined`;

exports[`Screen test Snapshot with loading as true  5`] = `undefined`;
